// ember-template-tag/dist/esm/parse-templates.js
import traverse from '@babel/traverse';
import { TEMPLATE_TAG_NAME, } from './util';
import parse from './template-parser';
export function isTemplateLiteralMatch(template) {
    return template.type === 'template-literal';
}
export const DEFAULT_PARSE_TEMPLATES_OPTIONS = {
    templateTag: TEMPLATE_TAG_NAME
};
/**
 * Parses a template to find all possible valid matches for an embedded template.
 * Supported syntaxes are template literals:
 *
 *   hbs`Hello, world!`
 *
 * And template tags
 *
 *   <template></template>
 *
 * The parser excludes any values found within strings recursively, and also
 * excludes any string literals with dynamic segments (e.g `${}`) since these
 * cannot be valid templates.
 *
 * @param template The template to parse
 * @param relativePath Relative file path for the template (for errors)
 * @param options optional configuration options for how to parse templates
 * @returns
 */
export function parseTemplates(template, relativePath, options = DEFAULT_PARSE_TEMPLATES_OPTIONS) {
    options.templateTag = options.templateTag || TEMPLATE_TAG_NAME;
    const ast = parse(template, options);
    return parseTemplatesFromAst(ast);
}
export function parseTemplatesFromAst(ast) {
    const results = [];
    traverse(ast, {
        // @ts-ignore
        enter(path) {
            const node = path.node;
            if (node.type === 'EmberTemplate') {
                results.push({
                    type: 'template-tag',
                    tagName: node.tagName,
                    contents: node.contentNode.quasis[0].value.raw,
                    contentRange: node.contentNode.range,
                    range: node.range,
                    startRange: {
                        start: node.startRange[0],
                        end: node.startRange[1],
                    },
                    endRange: { start: node.endRange[0], end: node.endRange[1] },
                });
            }
        }
    });
    return results;
}
